datalad/local/copy_file.py

```diff
@@ -0,0 +1,665 @@
+            code_py="""\
+            copy_file(dataset='path/to/myds', target_dir='path/to/myds',
+                      specs_from='path/to/specfile')""",
+            code_cmd="""\
+            datalad copy-file -d path/to/myds -t path/to/myds --specs-from specfile"""
+        ),
+        dict(
+            text="Read a specification from stdin and pipe the output of a find"
+                 " command into the copy-file command",
+            code_cmd="""\
+            find <expr> | datalad copy-file -d path/to/myds -t path/to/myds --specs-from -"""
+        )
+    ]
+
+    @staticmethod
+    @datasetmethod(name='copy_file')
+    @eval_results
+    def __call__(
+            path=None,
+            dataset=None,
+            recursive=False,
+            target_dir=None,
+            specs_from=None,
+            message=None):
+        # Concept
+        #
+        # Loosely model after the POSIX cp command
+        #
+        # 1. Determine the target of the copy operation, and its associated
+        #    dataset
+        #
+        # 2. for each source: determine source dataset, query for metadata, put
+        #    into target dataset
+        #
+        # Instead of sifting and sorting through input args, process them one
+        # by one sequentially. Utilize lookup caching to make things faster,
+        # instead of making the procedure itself more complicated.
+
+        if path and specs_from:
+            raise ValueError(
+                "Path argument(s) AND a specs-from specified, "
+                "this is not supported.")
+
+        ds = None
+        if dataset:
+            ds = require_dataset(dataset, check_installed=True,
+                                 purpose='copying into')
+
+        if target_dir:
+            target_dir = resolve_path(target_dir, dataset)
+        # TODO should we use ds.pathobj as target_dir, if none is given
+
+        if path:
+            # turn into list of absolute paths
+            paths = [resolve_path(p, dataset) for p in assure_list(path)]
+
+            # we already checked that there are no specs_from
+            if not target_dir:
+                if len(paths) == 1:
+                    raise ValueError("No target directory was given.")
+                elif len(paths) == 2:
+                    # single source+dest combo
+                    if paths[-1].is_dir():
+                        # check if we need to set target_dir, in case dest
+                        # is a dir
+                        target_dir = paths.pop(-1)
+                    else:
+                        specs_from = [paths]
+                else:
+                    target_dir = paths.pop(-1)
+
+            if not specs_from:
+                # in all other cases we have a plain source list
+                specs_from = paths
+
+        if not specs_from:
+            raise ValueError("Neither `paths` nor `specs_from` given.")
+
+        res_kwargs = dict(
+            action='copy',
```

Comment:
I think this should be "copy_file" to (1) be consistent with the module name and (2) to avoid a conflict with `push`, which already uses a "copy" action for something else.

Reply:
Done.